dune-project

```diff
@@ -1,2 +1,4 @@
-(lang dune 1.11)
+(lang dune 2.0)
 (name opam)
+
+(implicit_transitive_deps false)
```

Comment:
I'm not entirely sure this is such a good idea given we'd have to be careful not to depend on something non-abstract that we re-export the type of (e.g. Result.t when depending on the result library).
This is a good idea in theory but in practice this is a bit finicky to handle correctly.

Reply:
That's what I'm after, though - if we suddenly started using `Result`, then I'd prefer CI to fail immediately on 4.07 and reveal that the libraries are lacking an explicit dependency on the result library, rather than it "magically" working because it picked up the `Result` module via a dependency which depends on the `result` package (and which could choose in the future to stop depending on it, creating a very strange failure case).

As far as I can tell (from having just checked) the `implicit_transitive_deps` setting is not exported by the project. i.e. us adopting `implicit_transitive_deps` does not affect any users of the opam libraries.

We could have it set as `true` in the committed file and override in CI as a test, but I'm not (yet) convinced that the potential behaviour you're worried about would be a bad thing?

Reply:
> As far as I can tell (from having just checked) the implicit_transitive_deps setting is not exported by the project. i.e. us adopting implicit_transitive_deps does not affect any users of the opam libraries.

Are you sure? The setting is exported in the dune-package file.

> We could have it set as true in the committed file and override in CI as a test, but I'm not (yet) convinced that the potential behaviour you're worried about would be a bad thing?

That could be a sensible middle point. I remember some compiler devs horrified about `(implicit_transitive_deps false)` who basically said this should be a static check in dune instead, not a change to the compilation pass.
I like `(implicit_transitive_deps false)` on paper but i saw it bite back in opam-repo-ci a couple of times so I understand what they were saying.

> That's what I'm after, though - if we suddenly started using Result, then I'd prefer CI to fail immediately on 4.07 and reveal that the libraries are lacking an explicit dependency on the result library, rather than it "magically" working because it picked up the Result module via a dependency which depends on the result package (and which could choose in the future to stop depending on it, creating a very strange failure case).

Yes but the opposite is true for users. Let's say we have an .mli file with `val f : unit -> (string, string) result` and for whatever reason we switch the `result` type to `Result.t` (while depending on the `result` library) then the users who were using this function but not the `result` library will now fail if they tried to use it like this for instance:
```
match OpamModule.f () with
| Ok () -> fmt "yay"
| Error _ -> fmt "nay"
```
because now the type returned by `f` is abstract.

Reply:
mmh, sorry I tested again and it's not a problem for the opam library users but it's still a potential issue for the libraries we are using. We can always add more dependencies but for example in the case of `result` it would be nice to not depend on it when the library is not using it anymore.